src/agent/Agent.cpp

```diff
@@ -0,0 +1,222 @@
+#include "Agent.h"
+
+Agent::Agent(zsock_t* pipe) {
+    // TODO: remove connect() from here
+    redis_discover_.connect();
+
+    pipe_ = pipe;
+    request_ = NULL;
+    router_ = zsock_new(ZMQ_ROUTER);
+    poller_ = zpoller_new(pipe_);
+    zpoller_add(poller_, router_);
+    servers_ = zhash_new();
+
+    // update server list asap when Agent starts
+    next_discovery_ = zclock_mono();
+}
+
+Agent::~Agent() {
+    zsock_destroy(&router_);
+    zhash_destroy(&servers_);
+    zmsg_destroy(&request_);
+    zpoller_destroy(&poller_);
+}
+
+void Agent::unlock_servers() {
+    Server* item = (Server*) zhash_first(servers_);
+    while(item) {
+        item->busy_ = false;
+        item = (Server*) zhash_next(servers_);
+    }
+}
+
+// function to free the memory when zhash element is destroyed
+static void s_server_free (void *argument) {
+    Server* server = (Server *) argument;
+    delete server;
+}
+
+void Agent::process_control() {
+    zmsg_t *msg = zmsg_recv(pipe_);
+    char *command = zmsg_popstr(msg);
+
+    // Requester asks to send a message
+    if (streq(command, "REQUEST")) {
+        if (request_ != NULL) return;
+        request_ = msg;
+        msg = NULL;
+        expires_ = zclock_mono() + REQUEST_TIMEOUT;
+    }
+    free(command);
+    zmsg_destroy(&msg);
+}
+
+void Agent::discover_servers() {
+    if (zclock_mono() < next_discovery_) {
+        // too early to discover, skip
+        return;
+    }
+
+    // TODO: this is a quick and not efficient implementation, 
+    // need improvement in RedisDiscovery. Will do if will have time.
+    Server* server;
+    std::vector<std::string> all_tasks = redis_discover_.getTasks();
+    for (const std::string& task_name : all_tasks) {
+        std::vector<std::string> responders = redis_discover_.getAddresses(task_name);
+
+        for (const std::string& endpoint: responders) {
+            server = (Server*) zhash_lookup(servers_, endpoint.c_str());
+            if (server == NULL) {
+                int rc = zsock_connect(router_, endpoint.c_str());
+
+                // if connection failed, skip this server
+                if (rc != 0) continue;
+
+                // TODO: use zmonitor to check if enpoint is connected.
+                // need to add a small sleep so new connection can appear in socket
+                zclock_sleep(5);
```

Comment:
is this a race condition? what happens if 5 (ms?) is not enough?

Reply:
Yes, this is 5ms. After `zmq_connect` is OK (rc==0, if not mistaken), `zmq_socket` needs a small amount of time (do not know the exact number) so a new connection can be used. 

I don't like this `sleep` statement, but if I will not use it the request will be sent faster then a new connection will become available. Because I am using an async version of sockets (pair of router-router, ZeroMQ specific stuff), the socket will drop a message if it has nothing to send to. 

I would suppose, 5ms is actually too much because I tested it with the lower values. Probably, under high load, this sleep time should increase.

Reply:
If 5ms is not enough, then if the next message is sent to this exact connection and connection will not appear before the `zmsg_send`, a message will be dropped. I think `zmsg_send()` should return an error (c style, -1) in this situation and it is possible to check and throw an exception for the user.